sessions/2019/january/Session_22.md

```diff
@@ -0,0 +1,135 @@
+# SOFTWARE ENGINEERING IN JAVA
+
+## Session 22 (15/01/2019)
+
+- Java Memory Handling
+- Fixing Logger implmeentation
+- Sprint workflow in GitHub project for estimate effort
+- Lint checker
+- Microservices resources
+- Git advices
+
+### Notes
+
+#### Java memory
+
+![Java memory](https://cdncontribute.geeksforgeeks.org/wp-content/uploads/jvm-3.jpg "Java memory")
+
+Java uses their JVM to distribute its own memory use.
+
+In `Heap` is the zone where variables are. Pointers are in `JVM Language Stacks`.
+`PC Registers` is where CPU instructions are stored. `Native Method Stack` is where the methods that Java has built in are, and `Method Area` is self descriptive.
+
+`Class Loader` is where the compiled classes are found when the program is executed. As an example of this, I could to compile a program, and after it, proceed to remove the `.class` file from the `Class Loader` corresponding folder, after that, if I try to run the program will not do it, because the code isn't where is supossed to be, but if I compile again the code, this will not fail and work since the class is recreated into the `Class Loader` zone.
+
+#### Fixing Logger implementation
+
+Several fixes were made to the DefaultLogger class:
+
+```java
+@Component
+public class DefaultLogger implements Logger {
+    private final String ip;
+
+    DefaultLogger () {
+        InetAddress inetAddress;
+
+        try {
+            inetAddress = InetAddress.getLocalHost();
+        } catch (UnknownHostException e) {
+            inetAddress = null;
+        }
+
+        if (inetAddress != null){
+            ip = inetAddress.toString();
+        } else {
+            ip = "";
+        }
+    }
+
+    @Override
+    public void log(String message) {
+        Timestamp timestamp = new Timestamp(System.currentTimeMillis());
+        long threadId = Thread.currentThread().getId();
+
+        message = timestamp + "  INFO [" + threadId + "] " + ip + ": " + message;
+        System.out.println(message);
+    }
+}
+```
+
+A very important point here is that the system is asking for the IP address (`InetAddress.getLocalHost()`) in the constructor, so is executed once and the value is cached and used wherever is needed and not consulted each time when the method `log()` is called. Also, `null` inspection was made when the IP address of the system running cannot be loaded for some reason, avoiding to be using methods of `null` objects and receive `Null Pointer` exceptions
+
+By other side, `threads` values cannot be cached and they need to be consulted for each `request` made (each request is a `thread`).
```

Comment:
Different threads handle concurrent requests, but because threads aren't cheap to create Java does re-use a thread, after it has completed the execution of a request.

I think it's more accurate to say that we find out which thread is running the log call for each invocation of `log`.

Reply:
This sounds ok?

> By other side, `threads` values cannot be cached and they need to be consulted for each `request` made.  So we find out which thread is running the log call for each invocation of `log(String)`  method.
> 
> **NOTE:** Java application is a single process that have many threads. Different threads handle concurrent requests, but because threads aren't cheap to create, Java does re-use a thread, after it has completed the execution of a request.

Reply:
Sorry, I just discovered this!

```
On the other hand, per-threads values cannot be cached and they need to be calculated for each request made. So we find out which thread is running the log call for each invocation of log(String) method.

NOTE: A Java application is generally a single process that can have many threads. Different threads handle concurrent requests, but because threads aren't cheap to create, Spring does re-use a thread, after it has completed the execution of a request.
```